### PR TITLE
migrations + backend + sdk + frontend (combo pr for testing)

### PR DESCRIPTION
**Pull request by gtarpenning**
TODO: 
* frontend
     - refresh calls page on single call delete

Screenshot 2024-05-13 at 4 54 23 PM
Screenshot 2024-05-13 at 4 56 36 PM
Screenshot 2024-05-13 at 4 56 42 PM

delete-bulk-ham

### CODE REVIEW

**Comment by Yangyi-wandb** [comment]
Looks great!! 
In the delete modal: could we update the primary button text from 'Confirm' to 'Delete call'/ 'Delete calls' to make it more explicit? Also for 'Cancel' button, could we use ghost button style instead of primary (teal)? 
We would also need a checkbox for the trace column to select and deselect all, but it might be a different ticket. 
Screen Shot 2024-05-13 at 6 42 45 PM

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts [hunk 1]
```diff
@@ -202,12 +203,18 @@ const useCalls = (
...
   const getTsClient = useGetTraceServerClientContext();
   const loadingRef = useRef(false);
   const [callRes, setCallRes] =
     useState<traceServerClient.TraceCallsQueryRes | null>(null);
   const deepFilter = useDeepMemo(filter);
+
+  const [doRefetch, setDoRefetch] = useState(false);
+  const refetch = useCallback(() => {
+    setDoRefetch(r => !r);
+  }, []);
```
Comment (gtarpenning) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:216]: this is a hack!

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts [hunk 2]
```diff
@@ -270,11 +279,29 @@ const useCalls = (
...
 
+const useCallsDelete = () => {
+  const getTsClient = useGetTraceServerClientContext();
+
+  const callsDelete = useCallback(
+    (projectID, ids): Promise<traceServerClient.TraceCallsDeleteRes> => {
+      return getTsClient()
+        .callsDelete({project_id: projectID, ids})
+        .catch(e => {
+          console.error(e);
```
Comment (gtarpenning) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:296]: how do we want to handle errors? pass on to user?

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 3]
```diff
@@ -338,6 +372,14 @@ def ops_query(self, req: tsi.OpQueryReq) -> tsi.OpQueryRes:
         objs = [_ch_obj_to_obj_schema(call) for call in ch_objs]
         return tsi.OpQueryRes(op_objs=objs)
 
+    def ops_delete(self, req: tsi.OpsDeleteReq) -> tsi.OpsDeleteRes:
+        num_deleted = self._delete_objs(
+            req.project_id,
+            req.ids,
+            conditions=["is_op = 1"],
+        )
+        return tsi.OpsDeleteRes(success=req.ids == num_deleted)
```
Comment (gtarpenning) [inline weave/trace_server/clickhouse_trace_server_batched.py:381]: TODO: move into separate PR

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 4]
```diff
@@ -411,6 +453,33 @@ def objs_query(self, req: tsi.ObjQueryReq) -> tsi.ObjQueryRes:
...
+
+        summary = self._insert(
+            "object_versions",
+            data=[
+                (project_id, object.object_id, datetime.datetime.now())
+                for object in objects
+            ],
+            column_names=["project_id", "object_id", "deleted_at"],
+        )
+        return summary.written_rows
```
Comment (gtarpenning) [inline weave/trace_server/clickhouse_trace_server_batched.py:481]: TODO: move into next pr

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 5]
```diff
@@ -1293,3 +1364,32 @@ def _combine_conditions(conditions: typing.List[str], operator: str) -> str:
         raise ValueError(f"Invalid operator: {operator}")
     combined = f" {operator} ".join([f"({c})" for c in conditions])
     return f"({combined})"
+
+
+def find_call_descendants_in_memory_dfs(
```
Comment (gtarpenning) [inline weave/trace_server/clickhouse_trace_server_batched.py:1369]:
```suggestion
def find_call_descendants_useing_in_memory_dfs(
```

File: weave/trace_server/remote_http_trace_server.py [hunk 6]
```diff
@@ -167,6 +174,13 @@ def ops_query(
     ) -> tsi.OpQueryRes:
         return self._generic_request("/ops/query", req, tsi.OpQueryReq, tsi.OpQueryRes)
 
+    def ops_delete(
+        self, req: t.Union[tsi.OpsDeleteReq, t.Dict[str, t.Any]]
+    ) -> tsi.OpsDeleteRes:
+        return self._generic_request(
+            "/ops/delete", req, tsi.OpsDeleteReq, tsi.OpsDeleteRes
+        )
```
Comment (gtarpenning) [inline weave/trace_server/remote_http_trace_server.py:182]: TODO: move to next pr

File: weave/trace_server/sqlite_trace_server.py [hunk 7]
```diff
@@ -358,6 +406,9 @@ def op_read(self, req: tsi.OpReadReq) -> tsi.OpReadRes:
     def ops_query(self, req: tsi.OpQueryReq) -> tsi.OpQueryRes:
         raise NotImplementedError()
 
+    def ops_delete(self, req: tsi.OpsDeleteReq) -> tsi.OpsDeleteRes:
+        raise NotImplementedError()
```
Comment (gtarpenning) [inline weave/trace_server/sqlite_trace_server.py:410]: TODO: move

File: weave/trace_server/trace_server_interface.py [hunk 8]
```diff
@@ -199,6 +211,15 @@ class OpQueryRes(BaseModel):
     op_objs: typing.List[ObjSchema]
 
 
+class OpsDeleteReq(BaseModel):
+    project_id: str
+    ids: typing.List[str]
+
+
+class OpsDeleteRes(BaseModel):
+    success: bool
```
Comment (gtarpenning) [inline weave/trace_server/trace_server_interface.py:220]: MOVE

File: weave/trace_server/trace_server_interface.py [hunk 9]
```diff
@@ -323,6 +348,10 @@ def op_read(self, req: OpReadReq) -> OpReadRes:
     def ops_query(self, req: OpQueryReq) -> OpQueryRes:
         ...
 
+    @abc.abstractmethod
+    def ops_delete(self, req: OpsDeleteReq) -> OpsDeleteRes:
```
Comment (gtarpenning) [inline weave/trace_server/trace_server_interface.py:352]: MOVE

File: weave-js/src/components/PagePanelComponents/Home/Browse2/RunsTable.tsx [hunk 10]
```diff
@@ -490,14 +500,29 @@ export const RunsTable: FC<{
...
           return (
-            <CallLink
-              entityName={params.entity}
-              projectName={params.project}
-              opName={opVersionRefOpName(opVersion)}
-              callId={rowParams.row.id}
-              fullWidth={true}
-              preservePath={preservePath}
-            />
+            <>
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse2/RunsTable.tsx:503]: curious if you tried to use the DataGrid's selection mechanic

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/OverflowMenu.tsx [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+  setConfirmDelete: (confirmDelete: boolean) => void;
+  setSelectedCalls?: Dispatch<SetStateAction<CallSchema[]>>;
+  refetch?: () => void;
+}> = ({calls, confirmDelete, setConfirmDelete, setSelectedCalls, refetch}) => {
+  const {useCallsDelete} = useWFHooks();
+  const callsDelete = useCallsDelete();
+  const closePeek = useClosePeek();
+
+  if (new Set(calls.map(c => c.entity)).size > 1) {
+    throw new Error('Cannot delete calls from multiple entities');
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/OverflowMenu.tsx:82]: hmmm - is there a less destructive way to handle this?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts (hunk 2 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:296]: I think passing the Promise error to the caller is the best thing here for now.

File: weave-js/src/components/PagePanelComponents/Home/Browse2/RunsTable.tsx [hunk 12]
```diff
@@ -490,14 +500,29 @@ export const RunsTable: FC<{
...
+              <Checkbox
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse2/RunsTable.tsx:504]: I am wondering if we always want to show this. Seems like a lot of attention to draw and a super valuable real estate to takeup for an action that is an edge case. Something we should discuss with design tomorrow maybe?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/traceServerClient.ts [hunk 13]
```diff
@@ -78,6 +78,15 @@ export type TraceCallsQueryRes = {
   calls: TraceCallSchema[];
 };
 
+export type TraceCallsDeleteReq = {
+  project_id: string;
+  ids: string[];
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/traceServerClient.ts:83]: nit: prefer qualified id names: `call_ids`

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/traceServerClient.ts [hunk 14]
```diff
@@ -78,6 +78,15 @@ export type TraceCallsQueryRes = {
...
+};
+
+export type TraceCallsDeleteRes = {
+  success: boolean;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/traceServerClient.ts:87]: is there ever a case where a 200 response is not success?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts (hunk 1 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:216]: while this is a hack, you have the right idea. The implementation is a bit odd. Perhaps consider having the state be a timestamp, and refreshing sets the timestamp to the current timestamp.

Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:216]: In the future, we will have polling and better mechanics here

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts [hunk 15]
```diff
@@ -239,13 +246,14 @@ const useCalls = (
       setCallRes({calls: []});
     };
     getTsClient().callsSteamQuery(req).then(onSuccess).catch(onError);
-  }, [entity, project, deepFilter, limit, opts?.skip, getTsClient]);
+  }, [entity, project, deepFilter, limit, opts?.skip, doRefetch, getTsClient]);
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:249]: nit: awkward to have a symbol not used in the effect itself be in the deps list. This is because many linters check for this and it can be removed without much caution. Consider using the variable to ensure it is always included

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts (hunk 2 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:296]: basically just don't catch and you are good

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 16]
```diff
@@ -302,6 +305,37 @@ def calls_query_stream(
...
+            root_ids=req.ids,
+            all_calls=all_calls,
+        )
+
+        # create insert payload for all descendants (and parents)
+        data = [(req.project_id, d, deleted_at) for d in all_descendants]
+        summary = self._insert(
+            "call_parts", data=data, column_names=["project_id", "id", "deleted_at"]
+        )
+        return tsi.CallsDeleteRes(success=summary.written_rows == len(all_descendants))
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:337]: this would be very odd to be false - did you see that happen?

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 17]
```diff
@@ -302,6 +305,37 @@ def calls_query_stream(
                 _ch_call_dict_to_call_schema_dict(ch_dict)
             )
 
+    def calls_delete(self, req: tsi.CallsDeleteReq) -> tsi.CallsDeleteRes:
+        deleted_at = datetime.datetime.now()
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:309]: might want to just add a sanity check of limiting to 100 ids at a time

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 18]
```diff
@@ -1293,3 +1329,32 @@ def _combine_conditions(conditions: typing.List[str], operator: str) -> str:
         raise ValueError(f"Invalid operator: {operator}")
     combined = f" {operator} ".join([f"({c})" for c in conditions])
     return f"({combined})"
+
+
+def find_call_descendants_in_memory_dfs(
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:1334]: nit: prefer to not leak implementation details (dfs) into the function name

Comment (tssweeney) [reply weave/trace_server/clickhouse_trace_server_batched.py:1334]: also, the "in memory" part is presumed since you are passing an in memory list of calls

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 19]
```diff
@@ -972,8 +1007,9 @@ def _select_objs_query(
                 version_count,
                 is_latest
             FROM object_versions_deduped
-            WHERE project_id = {{project_id: String}}
-            AND {conditions_part}
+            WHERE project_id = {{project_id: String}} AND
+                deleted_at IS NULL AND
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:1011]: just adding this in anticipation?? (:

File: weave/trace_server/migrations/002_add_deleted_at.down.sql [hunk 20]
```diff
@@ -0,0 +1,67 @@
+ALTER TABLE object_versions DROP COLUMN deleted_at;
```
Comment (tssweeney) [inline weave/trace_server/migrations/002_add_deleted_at.down.sql:1]: This is a hard file to parse given that it is mostly a copy. Can you please comment to make it clear what changed vs what is just the copy over?

File: weave/trace_server/trace_server_interface.py [hunk 21]
```diff
@@ -135,6 +138,15 @@ class CallReadRes(BaseModel):
     call: CallSchema
 
 
+class CallsDeleteReq(BaseModel):
+    project_id: str
+    ids: typing.List[str]
```
Comment (tssweeney) [inline weave/trace_server/trace_server_interface.py:143]: again, prefer `call_ids`

File: weave/trace_server/trace_server_interface.py [hunk 22]
```diff
@@ -135,6 +138,15 @@ class CallReadRes(BaseModel):
...
+
+
+class CallsDeleteRes(BaseModel):
+    success: bool
```
Comment (tssweeney) [inline weave/trace_server/trace_server_interface.py:147]: not sure how this can be non-true

File: weave/tests/test_weave_client.py [hunk 23]
```diff
@@ -239,6 +239,65 @@ def test_calls_query(client):
     client.finish_call(call0, None)
 
 
+def test_calls_delete(client):
```
Comment (tssweeney) [inline weave/tests/test_weave_client.py:242]: please verify that you ran these tests in clickhouse mode

File: weave/trace_server/clickhouse_trace_server_batched.py (hunk 19 above)
Comment (gtarpenning) [reply weave/trace_server/clickhouse_trace_server_batched.py:1011]: sorry yeah, i have another pr that this should be moved to ;)

File: weave/tests/test_weave_client.py (hunk 23 above)
Comment (gtarpenning) [reply weave/tests/test_weave_client.py:242]: @tssweeney im sure this is easy to do but for the life of me I can't figure out the magic incantation...
